R/zchunk_LA124.heat.R

```diff
@@ -40,70 +40,175 @@ module_energy_LA124.heat_DISABLED <- function(command, ...) {
+    A24.globaltech_coef[,as.character(HISTORICAL_YEARS[-1])] <- NA
+
+    # Interpolate to fill in missing globaltech_coef historical years
+    A24.globaltech_coef %>%
+      gather(year, value, -supplysector, -subsector, -technology, -minicam.energy.input) %>%
+      mutate(year = as.integer(year))  %>%
+      arrange(year) %>%
+      group_by(technology, subsector, supplysector, minicam.energy.input)  %>%
+      mutate(value = approx_fun(year, value)) %>%
+      left_join(distinct(calibrated_techs)) %>%
+      select(-calibration, -secondary.output) -> L124.globaltech_coef
+
+    # Heat output: fuel inputs to heat divided by exogenous input-output coefficients
+    L124.in_EJ_R_heat_F_Yh %>%
+      left_join(L124.globaltech_coef %>%
+                  rename(temp=value),
+                      by=c("sector", "fuel", "year")) %>%
+      mutate(value=value/temp)%>%
+      select(-temp) %>%
+      filter(GCAM_region_ID %in% heat_regionIDs)-> L124.out_EJ_R_heat_F_Yh
+
+    # Secondary output of heat from main activity CHP plants
+    L1011.en_bal_EJ_R_Si_Fi_Yh %>%
+      filter(sector == "out_electricity_heat" , GCAM_region_ID %in% heat_regionIDs) %>%
+      mutate(sector = sub("out_", "", sector)) %>%
+      left_join(enduse_fuel_aggregation, by="fuel")  %>%
+      select(-fuel, -electricity, -bld, -industry, -trn)%>%
+      rename(fuel=heat) %>%
+      gather(key=year, value=value, -fuel, -sector, -GCAM_region_ID) %>%
+      group_by(fuel, sector, GCAM_region_ID, year) %>%
+      summarise(value=sum(value)) %>%
+      filter(fuel != is.na(fuel)) %>%
+      mutate(year = sub("X", "", year)) %>%
+      mutate(year = as.integer(year)) -> L124.out_EJ_R_heatfromelec_F_Yh
+
+
+    #Secondary output coefficients on heat produced by main activity CHP plants
+    L1231.out_EJ_R_elec_F_tech_Yh %>%
+      filter(GCAM_region_ID %in% heat_regionIDs) %>%
+      filter(technology %in% calibrated_techs$technology[calibrated_techs$secondary.output == "heat"]) %>%
+      gather(key=year, value=value, -fuel, -sector, -GCAM_region_ID, -technology) %>%
+      mutate(year = sub("X", "", year)) %>%
+      mutate(year = as.integer(year)) %>%
+      left_join(L124.out_EJ_R_heatfromelec_F_Yh %>%
+                  rename(value_elec=value) %>%
+                  rename(temp=sector), by=c("GCAM_region_ID", "fuel", "year")) %>%
+      mutate(value=value_elec/value) %>%
+      select(-value_elec, -temp)-> L124.heatoutratio_R_elec_F_tech_Yh
+
+    #Reset missing and infinite values (applicable for CC in the base years) to 0
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      mutate(value=if_else(is.na(value)|is.infinite(value), 0, value))-> L124.heatoutratio_R_elec_F_tech_Yh
+
+    #Drop all rows where all years are 0
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      group_by(GCAM_region_ID, technology, sector, fuel) %>%
+      summarise(sum=sum(value)) %>%
+      filter(sum==0) -> all_years_0
+
+    L124.heatoutratio_R_elec_F_tech_Yh %>%
+      left_join(all_years_0) %>%
```

Comment:
Here and elsewhere, is it possible to use a more restrictive join, e.g. `left_join_error_no_match`?